proposed/2021/VulnerabilityInfoInPMUI.md

```diff
@@ -0,0 +1,52 @@
+* Track via telemetry when an action (installed, update, uninstall) is performed on a package with a vulnerability from within the Package Manager UI.
+
+## Non-Goals
+
+* Improve the performance of querying package vulnerability information.
+* Change the architecture of the PMUI to not require querying information for the Installed tab to know if there is a deprecated package or vulnerability for the warning icon, and once again for the packages list. The existing architecture of the PMUI will be used and existing patterns will be followed.
+* Make any improvements to the package deprecation experience.
+* Change the registration API to pull in more details about a package vulnerability. This will be considered in a future iteration.
+* Show a warning icon in the Solution Explorer when packages with vulnerabilities are installed. This will be considered in a future iteration.
+
+## Solution
+
+A list of vulnerabilities for a package is read from the server, along with other data, whenever a package version is selected (DetailControlModel.SetCurrentPackageAsync) or when the selection version changes (DetailControlModel.SelectedVersionChangedAsync). This data is stored in the DetailedPackageMetadata and is used by the Details Pane. The vulnerabilities information exposed in the PackageSolutionDetailControlModel which implements DetailedPackageMetadata is currently limited to whether a vulnerability exists and what the maximum severity of the vulnerability is. This will be changed to expose instead a list of vulnerabilities, their severity, and their advisory links.
+
+The Details Pane will have a new section in the view called "Vulnerabilities" that will be visible when a package has one or more vulnerabilities. The Vulnerabilities section will display the highest severity of the vulnerabilities and show a list of vulnerabilities, their severity, and a link to the advisory. To stay consistent with the current design of the details pane, the vulnerabilities section will be created as its own UserControl that will be used by the DetailControl.
+
+The PackageManagerTopPanel will be modified to handle showing a warning icon when there are either deprecated or installed packages, or both. A new method will be added to this class to handle getting the count of the vulnerable packages installed which will be used in the tooltip and also to determine whether or not to show the warning icon for vulnerabilities.
+
+The PackageItemViewModel will have a new method to query the Vulnerability status of a given package asynchronously, much like it currently reads package deprecation status. This information will be used, along with the deprecation status, to determine if the warning icon should appear by the package in the package list. The tooltip for the icon will be updated to display information about the vulnerabilities and/or deprecated packages.
```

Comment:
> The tooltip for the icon will be updated to display information about the vulnerabilities and/or deprecated packages.

How will this look? Is there a sample from a PM or your own mockup? I'd be interested to see the layout and understand what we are deeming as worthy of tooltip text. Eg, is the max severity only shown? Are all severities listed? How big can this tooltip become in the average/worst cases?

Reply:
@chgill-MSFT has a spec that he'll be adding to source control that will act as the source of truth for the look and design of the feature. If you take a look at the screenshot in the e-mail spec, it shows a good picture of what the tooltip will look like. I would rather not include that image in the technical spec as well, because if it changes in the design spec, we'll need to update it in two places. As soon as the design spec is up in the repo, I'll add a reference to it in this spec.